packages/core/src/lifecycle-registry.ts

```diff
@@ -0,0 +1,243 @@
+    > = new Map();
+    for (const binding of bindings) {
+      const group = this.getObserverGroup(binding);
+      let bindingsInGroup = groupMap.get(group);
+      if (bindingsInGroup == null) {
+        bindingsInGroup = [];
+        groupMap.set(group, bindingsInGroup);
+      }
+      bindingsInGroup.push(binding);
+    }
+    // Create an array for group entries
+    const groups: LifeCycleObserverGroup[] = [];
+    for (const [group, bindingsInGroup] of groupMap) {
+      groups.push({group, bindings: bindingsInGroup});
+    }
+    // Sort the groups
+    return groups.sort((g1, g2) => {
+      const i1 = this.options.groupsByOrder.indexOf(g1.group);
+      const i2 = this.options.groupsByOrder.indexOf(g2.group);
+      if (i1 !== -1 || i2 !== -1) {
+        // Honor the group order
+        return i1 - i2;
+      } else {
+        // Neither group is in the pre-defined order
+        // Use alphabetical order instead so that `1-group` is invoked before
+        // `2-group`
+        return g1.group < g2.group ? -1 : g1.group > g2.group ? 1 : 0;
+      }
+    });
+  }
+
+  /**
+   * Notify an observer group of the given event
+   * @param group A group of bindings for life cycle observers
+   * @param event Event name
+   */
+  protected async notifyObservers(
+    observers: LifeCycleObserver[],
+    bindings: Readonly<Binding<LifeCycleObserver>>[],
+    event: keyof LifeCycleObserver,
+  ) {
+    if (!this.options.parallel) {
+      let index = 0;
+      for (const observer of observers) {
+        debug(
+          'Invoking %s observer for binding %s',
+          event,
+          bindings[index].key,
+        );
+        index++;
+        await this.invokeObserver(observer, event);
+      }
+      return;
+    }
+
+    // Parallel invocation
+    const notifiers = observers.map((observer, index) => {
+      debug('Invoking %s observer for binding %s', event, bindings[index].key);
+      return this.invokeObserver(observer, event);
+    });
+    await Promise.all(notifiers);
+  }
+
+  /**
+   * Invoke an observer for the given event
+   * @param observer A life cycle observer
+   * @param event Event name
+   */
+  protected async invokeObserver(
+    observer: LifeCycleObserver,
+    event: keyof LifeCycleObserver,
+  ) {
+    if (typeof observer[event] === 'function') {
+      await observer[event]!();
+    }
+  }
+
+  /**
+   * Emit events to the observer groups
+   * @param events Event names
+   * @param groups Observer groups
+   */
+  protected async notifyGroups(
+    events: (keyof LifeCycleObserver)[],
+    groups: LifeCycleObserverGroup[],
+    reverse = false,
+  ) {
+    const observers = await this.observersView.values();
+    const bindings = this.observersView.bindings;
+    if (reverse) groups = groups.reverse();
+    for (const group of groups) {
+      const observersForGroup: LifeCycleObserver[] = [];
+      const bindingsInGroup = reverse
+        ? group.bindings.reverse()
```

Comment:
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reverse
> The `reverse()` method reverses an array in place. The first array element becomes the last, and the last array element becomes the first.

I don't think it's a good idea to modify `group.bindings` array.

```suggestion
        ? [...group.bindings].reverse()
```

Please add a test that fails in the current implementation and passes with the fix in place.